liquidity-rewards/contracts/UniswapPoolRewards.sol

```diff
@@ -1,5 +1,130 @@
+     * => rewardRate = 1.87 x 10e18 (token decimals) = 1870000000000000000
+     */
+    uint256 public rewardRate = 1870000000000000000;
+    uint256 public lastUpdateTime;
+    uint256 public rewardPerTokenStored;
+    mapping(address => uint256) public userRewardPerTokenPaid;
+    mapping(address => uint256) public rewards;
+
+    event Deposited(address indexed user, uint256 amount);
+    event Withdrawn(address indexed user, uint256 amount);
+    event RewardClaimed(address indexed user, uint256 reward);
+
+    constructor(address _elimuTokenAddress, address _lpTokenAddress) {
+        lpToken = IERC20(_lpTokenAddress);
+        elimuToken = IERC20(_elimuTokenAddress);
+    }
+
+
+    // Returns the amount of rewards that correspond to each staked token
+    function rewardPerToken() public view returns (uint256) {
+        if (totalSupply() == 0) {
+            return rewardPerTokenStored;
+        }
+        return
+            rewardPerTokenStored.add(
+                block.timestamp
+                    .sub(lastUpdateTime)
+                    .mul(rewardRate)
+                    .mul(1e18)
+                    .div(totalSupply())
+            );
+    }
+
+    // Returns the amount that an account can claim
+    function rewardsEarned(address account) public view returns (uint256) {
+        return
+            balanceOf(account)
+                .mul(rewardPerToken().sub(userRewardPerTokenPaid[account]))
+                .div(1e18)
+                .add(rewards[account]);
+    }
+
+    // deposit visibility is public as overriding LPTokenWrapper's deposit() function
+    function deposit(uint256 amount) public override {
+        require(amount > 0, "Cannot stake 0");
+        require(address(lpToken) != address(0), "Liquidity Pool Token has not been set yet");
+
+        _updateAccountReward(msg.sender);
+
+        super.deposit(amount);
+        emit Deposited(msg.sender, amount);
+    }
+
+    function withdraw(uint256 amount) public override{
+        require(amount > 0, "Cannot withdraw 0");
+        require(address(lpToken) != address(0), "Liquidity Pool Token has not been set yet");
+
+        _updateAccountReward(msg.sender);
+
+        super.withdraw(amount);
+        emit Withdrawn(msg.sender, amount);
+    }
+
+    // Shortcut to be able to unstake tokens and claim rewards in one transaction
+    function withdrawAndClaim() external {
+        withdraw(balanceOf(msg.sender));
+        claimReward();
+    }
+
+    function claimReward() public {
+        require(address(lpToken) != address(0), "Liquidity Pool Token has not been set yet");
+
+        _updateAccountReward(msg.sender);
+
+        uint256 reward = rewardsEarned(msg.sender);
+
+        require(reward > 0, "Nothing to claim");
+
+        rewards[msg.sender] = 0;
+        elimuToken.transfer(msg.sender, reward);
+        emit RewardClaimed(msg.sender, reward);
+    }
+
+    function depositReward(uint256 reward) external onlyOwner {
```

Comment:
Does this mean that only the owner of the smart contract will be allowed to deposit funds to it? So if some random Ethereum account tries to deposit `$ETH` or `$ELIMU` into the smart contract, it will not work?

Reply:
It will work!
Actually, this contract method could be deleted as our calculations no longer depend on the actual value of the $ELIMU stored on the contract.
Also, we may reach a case when there are no longer enough $ELIMU tokens on the contract to reward the lp users. It could lead to failed lp users' transactions but not actual loss of rewards.